### ISSUE TITLE
Support X11 Forwarding

### ISSUE THREAD

**Issue by bandogora** [PowerShell/Win32-OpenSSH#1515]
# Requesting X11 Forwarding Feature
Looking through the docs I see:
> The following configuration options are not available in the OpenSSH version that ships in Windows Server 2019 and Windows 10 1809:
>...
> X11DisplayOffset
> X11Forwarding
> X11UseLocalhost
> XAuthLocation

But I can't see why these features can't be implemented. I can successfully forward X11 using PuTTY or Ubuntu on WLS with Xming. So why can't I achieve this with OpenSSH on PowerShell?

X11 forwarding is incredibly important to my work and whenever I need it I find myself enabling WLS or switching to a Linux distro, which is pretty inconvenient when I need Windows. I'm sure I'm not the only user with this issue.

I've tried a few different configurations to get this working but I always get the same errors:
```
*WARNING* X Window Display Initialization failure
*WARNING* (DISPLAY "<not defined>")
```
Things I've tried:
```
set DISPLAY=localhost:0.0
set DISPLAY=127.0.0.1
ssh -Y example@192.168.1.1
ssh -X example@192.168.1.1
etc...
```

I've tried various other options as well like disabling X11 authentication as I've come back to try and solve this issue several times on my end.

If there is a way to configure OpenSSH to work with Xming or VcXsrv please add it to the docs.

**"OpenSSH for Windows" version**
8.0.0.0
 
**Server OperatingSystem**
 CentOS release 5.11 (Final)

**Client OperatingSystem**
Windows 10 Pro

**Comment by pd93** [comment]
@manojampalam mentioned support for this (or lack thereof) in #1438.

> Haven't explored X11 forwarding on Windows yet. Its currently not in scope. If you feel its an easy feature to add, please propose on how it could be done. A PR with changes would be helpful.

Our use-case for this is VSCode's Remote-SSH extension, which we'd like to use to connect to a remote development machine and then forward the X-Window of an Android Emulator back, so that we can do Flutter development remotely.

-- Edit: Also related: #1181

-- Edit 2: For anyone reading this that wants X11 for remote Flutter dev, we decided that using `ssh -R` or `RemoteForward` to forward the adb port to our remote machine was a much more effective solution. That being said, having the option of X11 is always nice.

**Comment by riverar** [comment]
So I was looking into implementing this and turns out, it already works in `OpenSSH_for_Windows_8.1p1` once you resolve the `ASKPASS` bug. Windows 10 currently ships with `OpenSSH_for_Windows_7.7p1`.

![Screenshot of X11 windows on Windows 10, using OpenSSH_..._8.1](https://user-images.githubusercontent.com/475132/74490139-74c1f180-4e7c-11ea-9943-8b8a6d23c13d.png)


@manojampalam Any updates on when a new copy will ship with Windows?

@pd93 If you grab newer binaries, your scenario should be unblocked!

**Comment by bagajjal** [comment]
@riverar - We are shipping OpenSSH V8.1 in next windows release that might be available in July 2020.

**Comment by riverar** [comment]
@bagajjal That's great news, thanks! We should try to get that `ASKPASS` issue fixed before it ships so users won't need that `C:\dev\tty` file on disk. I can submit a PR if desired.

**Comment by bagajjal** [comment]
Sure, please submit a PR.

**Comment by pd93** [comment]
@riverar Amazing. Thanks for the info. I've already manually updated v8.0 in order to get `ProxyJump` functionality, so it's not too bad to update to v8.1.

**Comment by riverar** [comment]
@pd93 You can probably keep 8.0. Just create a `C:\dev\tty` file to workaround an ASKPASS quirk.

**Comment by jasonzio** [comment]
@bagajjal can you confirm the 8.1 update made it into Windows 10 version 2004?

**Comment by bagajjal** [comment]
@jasonzio - v8.1 goes into the next release (not into win10 version 2004) available during fall time.

**Comment by riverar** [comment]
> @jasonzio - v8.1 goes into the next release (not into win10 version 2004) available during fall time.

Why is OpenSSH an optional FOD component if you're going to not deliver outside the OS schedule? Seems way too long wait for updates.

**Comment by pd93** [comment]
@bagajjal I've seen 3 different people say the same thing 3 times now - that the latest version of OpenSSH will make it into the "next" version of Windows. It never happens :/

I agree with @riverar. Microsoft really need to review their commitment to OpenSSH and begin to ship updates on a regular basis. It's simply ridiculous that something as fundamental as SSH only gets an updated every 2 years. Especially with so many outstanding bugs in the "current" version (7.9).

**Comment by bagajjal** [comment]
@riverar - We can't go the route of the windows store as ssh is been used in different windows SKUs that wouldn't get the update if we go down that route. 

@pd93 - you  need to understand the latest version will be part of next release that will be available in fall season. I have been consistent on the time line when the fix will be available. 

I agree there is a quite a delay in getting the new version but it's because of the resource allocation towards this project and windows release timelines. Going forward, I'm committed to get the latest version of openssh into every windows update.

**Comment by riverar** [comment]
@bagajjal I wasn't suggesting the Microsoft Store; was just wondering why it's a Feature On Demand / Optional Feature if it's just going to ship twice a year with the OS. But sounds like you covered the reason -- resource allocation. Thanks for the response!

**Comment by bagajjal** [comment]
@riverar - Twice a year update happened because of several reasons (like resource shuffling, project funding, priorities towards other projects, windows timeline). As I mentioned, I will make sure I ship the latest and greatest from the next windows release.

**Comment by pd93** [comment]
@bagajjal Apologies if I sounded too critical. Have just been slightly frustrated by the timeframes (or lack thereof). The replies and clarification here are greatly appreciated. Looking forward to testing it out later this year.

**Comment by kingyue737** [comment]
Hi @bagajjal , I'm also looking forward to the newest openssh which supports X11 forwarding. I've installed the preview of Windows10 20H2, but the version of openssh is still v7.7. Will the official release of 20H2 ship with openssh v8.1?

**Comment by bagajjal** [comment]
It's part of windows fall update which is yet to come. I see it's planned to release on 10/2/2020.

**Comment by zuboje** [comment]
> It's part of windows fall update which is yet to come. I see it's planned to release on 10/2/2020.

How do you setup X11 with Powershell core?

**Comment by zuboje** [comment]
> It's part of windows fall update which is yet to come. I see it's planned to release on 10/2/2020.

It is not delivered in fall update or maybe your "fall update" means some other update:

![image](https://user-images.githubusercontent.com/5181730/97462518-0c2c7f00-18fc-11eb-9d65-b548ad139534.png)

**Comment by pd93** [comment]
@zuboje Please see [the pinned issue](https://github.com/PowerShell/Win32-OpenSSH/issues/1693) logged by a MSFT employee. This explains why an update OpenSSH was not included in 20H2

**Comment by kingyue737** [comment]
> So I was looking into implementing this and turns out, it already works in `OpenSSH_for_Windows_8.1p1` once you resolve the `ASKPASS` bug. Windows 10 currently ships with `OpenSSH_for_Windows_7.7p1`.
> 
> ![Screenshot of X11 windows on Windows 10, using OpenSSH_..._8.1](https://user-images.githubusercontent.com/475132/74490139-74c1f180-4e7c-11ea-9943-8b8a6d23c13d.png)
> 
> @manojampalam Any updates on when a new copy will ship with Windows?
> 
> @pd93 If you grab newer binaries, your scenario should be unblocked!

Hi @riverar , I've installed openssh v8.1 and VcXsrv on my windows10 pro.  But it still shows can't open display. Can you give a simple example how to use x11 forwarding with openssh and powershell?

**Comment by dhruvinsh** [comment]
I got it working with my win 10 pc 
```
Edition	Windows 10 Pro
Version	21H1
Installed on	‎5/‎12/‎2021
OS build	19043.1202
Experience	Windows Feature Experience Pack 120.2212.3530.0
```

if in `cmd.exe`, do this:
```
set DISPLAY=localhost:0.0
ssh -Y host
> ~ xclock
```

if in `powershell` do this:
```
$env:DISPLAY='localhost:0.0'
ssh -Y host
> ~ xclock
```

still need to figure out how will it work in `Microsoft Terminal`.

**Comment by uchenily** [comment]
On Microsoft Terminal, we can set like this:
```shell
PowerShell -Command "$env:DISPLAY='localhost:0.0'; ssh.exe -Y 192.168.182.128"
```
Besides, we need to run an additional x11-server program, I found MobaXterm work well.(We must install MobaXterm first. If there is a better way, please tell me.)

```shell
C:\Users\${USERNAME}\Documents\MobaXterm\slash\bin\XWin_MobaX.exe -silent-dup-error -notrayicon -nolisten inet6 -hostintitle +bs -clipboard -nowgl -multiwindow -noreset :0
```

![图片](https://user-images.githubusercontent.com/39831719/136685894-0fcc0c74-2156-4b25-90bb-9bbaeb48e852.png)
